File: app/core/derived/edge.py
"""Edge (reference) and node_link registry derivation from inline node content."""

from __future__ import annotations

import json
import re
import sqlite3
import uuid as uuid_module
from typing import Any

from app.core.operation import Operation
from app.core.uuid import uuidv7


# Namespace for deterministic link UUIDs generated for legacy bare-target links
# until the one-time operation-log migration assigns stable UUIDs.
_LEGACY_LINK_UUID_NAMESPACE = uuid_module.UUID("0194a1b2-3c4d-5e6f-7a8b-9c0d1e2f3a4b")


class NodeLinkInstance:
    """A single inline link instance extracted from an AST."""

    def __init__(
        self,
        link_uuid: str,
        target_uuid: str,
        ref_type: str,
        label: str | None,
    ) -> None:
        self.link_uuid = link_uuid
        self.target_uuid = target_uuid
        self.ref_type = ref_type
        self.label = label


def _link_uuid_for_legacy_target(source_id: str, target_id: str) -> str:
    """Return a deterministic link UUID for a legacy bare-target link.

    The one-time migration in ``app/db/migrations/normalize_node_link_uuids.py``
    rewrites operation-log payloads to include stable link UUIDs. Until it runs,
    derived-state rebuilds must still produce deterministic node_link rows so
    counts and backlinks remain stable across rebuilds.
    """
    return str(uuid_module.uuid5(_LEGACY_LINK_UUID_NAMESPACE, f"{source_id}:{target_id}"))


def extract_node_link_instances(
    content: list[dict[str, Any]], source_id: str
) -> list[NodeLinkInstance]:
    """Return all inline link instances from the content AST.

    Supports ``node_link`` pills, legacy ``ref`` nodes, and raw ``[[uuid]]`` text.
    Legacy bare ``targetUuid`` identifiers are assigned a deterministic link UUID.
    """
    instances: list[NodeLinkInstance] = []
    seen_legacy_targets: set[str] = set()

    def _visit(node: Any) -> None:
        if not isinstance(node, dict):
            return
        ctype = node.get("type")
        if ctype == "node_link" and node.get("link_id"):
            link_id = str(node["link_id"])
            parts = link_id.split(":", 1)
            target_id = parts[0]
            link_uuid = parts[1] if len(parts) > 1 else _link_uuid_for_legacy_target(source_id, target_id)
            if target_id:
                instances.append(
                    NodeLinkInstance(
                        link_uuid=link_uuid,
                        target_uuid=target_id,
                        ref_type=node.get("ref_type", "node"),
                        label=node.get("label"),
                    )
                )
        elif ctype == "ref" and node.get("targetId"):
            target_id = str(node["targetId"])
            link_uuid = _link_uuid_for_legacy_target(source_id, target_id)
            if target_id not in seen_legacy_targets:
                seen_legacy_targets.add(target_id)
                instances.append(
                    NodeLinkInstance(
                        link_uuid=link_uuid,
                        target_uuid=target_id,
                        ref_type="node",
                        label=node.get("label"),
                    )
                )
        elif ctype == "text" and isinstance(node.get("text"), str):
            text = node["text"]
            for match in re.finditer(r"\[\[([^\]]+)\]\]", text):
                target_id = match.group(1)
                link_uuid = _link_uuid_for_legacy_target(source_id, target_id)
                if target_id not in seen_legacy_targets:
                    seen_legacy_targets.add(target_id)
                    instances.append(
                        NodeLinkInstance(
                            link_uuid=link_uuid,
                            target_uuid=target_id,
                            ref_type="node",
                            label=None,
                        )
                    )
        children = node.get("children")
        if isinstance(children, list):
            for child in children:
                _visit(child)

    for child in content:
        _visit(child)
    return instances


def rebuild_node_links_for_node(conn: sqlite3.Connection, op: Operation) -> None:
    """Synchronise ``node_link`` rows for ``op.payload.nodeId``."""
    node_id = op.payload["nodeId"]
    workspace_id = op.envelope.workspace_id
    row = conn.execute("SELECT content FROM node WHERE id = ?", (node_id,)).fetchone()
    if row is None:
        return
    content = json.loads(row[0])
    desired = extract_node_link_instances(content, node_id)
    desired_ids = {link.link_uuid for link in desired}

    ts = op.envelope.timestamp.isoformat() if op.envelope.timestamp else None

    for link in desired:
        conn.execute(
            """
            INSERT INTO node_link (
                id, workspace_id, source_id, target_id, type, label,
                click_count, last_navigated_at, created_at, updated_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            ON CONFLICT(id) DO UPDATE SET
                source_id = excluded.source_id,
                target_id = excluded.target_id,
                type = excluded.type,
                label = excluded.label,
                updated_at = excluded.updated_at
            """,
            (
                link.link_uuid,
                workspace_id,
                node_id,
                link.target_uuid,
                link.ref_type,
                link.label,
                0,
                None,
                ts,
                ts,
            ),
        )

    for stale_id in _existing_node_link_ids(conn, node_id) - desired_ids:
        conn.execute("DELETE FROM node_link WHERE id = ?", (stale_id,))


def _existing_node_link_ids(conn: sqlite3.Connection, source_id: str) -> set[str]:
    rows = conn.execute("SELECT id FROM node_link WHERE source_id = ?", (source_id,)).fetchall()
    return {row["id"] for row in rows}


def rebuild_edges_for_node(conn: sqlite3.Connection, op: Operation) -> None:
    """Synchronise ``edge`` rows of type ``reference`` for ``op.payload.nodeId``.

    ``edge`` is a deduplicated graph projection derived from ``node_link``.
    It exists for backwards compatibility until all callers migrate to
    ``node_link``.
    """
    node_id = op.payload["nodeId"]
    workspace_id = op.envelope.workspace_id
    rebuild_node_links_for_node(conn, op)

    desired_edges: dict[tuple[str, str], str | None] = {}
    for row in conn.execute(
        "SELECT target_id, label FROM node_link WHERE source_id = ?",
        (node_id,),
    ).fetchall():
        key = (node_id, row["target_id"])
        if key not in desired_edges:
            desired_edges[key] = row["label"]

    existing = conn.execute(
        "SELECT id, target_id, metadata FROM edge WHERE source_id = ? AND type = ?",
        (node_id, "reference"),
    ).fetchall()
    existing_targets = {row["target_id"]: row for row in existing}

    ts = op.envelope.timestamp.isoformat() if op.envelope.timestamp else None

    for (_, target_id), label in desired_edges.items():
        if target_id in existing_targets:
            continue
        conn.execute(
            """
            INSERT INTO edge (
                id, workspace_id, source_id, target_id, type,
                property_schema_id, metadata, created_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                uuidv7(),
                workspace_id,
                node_id,
                target_id,
                "reference",
                None,
                json.dumps({"label": label}),
                ts,
            ),
        )

    desired_target_ids = {target_id for (_, target_id) in desired_edges}
    for target_id, row in existing_targets.items():
        if target_id not in desired_target_ids:
            conn.execute("DELETE FROM edge WHERE id = ?", (row["id"],))

File: frontend/src/core/derived/edge.ts
import { type Database } from 'sql.js';
import { uuidv7 } from '../uuid';
import { queryAll, queryOne } from '../db/sqlite';
import { parseLinkId } from '@/lib/astBuilder';
import { generateLegacyLinkUuid } from '@/utils/uuid';

const REF_MARK_RE = /\[\[([^\]]+)\]\]/g;

interface NodeLinkInstance {
  linkUuid: string;
  targetUuid: string;
  type: string;
  label: string | null;
}

function extractNodeLinkInstances(content: unknown[], sourceId: string): NodeLinkInstance[] {
  const instances: NodeLinkInstance[] = [];
  const seenLegacyTargets = new Set<string>();

  function walk(node: unknown) {
    const c = node as {
      type?: string;
      targetId?: string;
      label?: string;
      text?: string;
      link_id?: string;
      ref_type?: string;
      children?: unknown[];
    };
    if (c.type === 'node_link' && c.link_id) {
      const { nodeUuid, linkUuid } = parseLinkId(c.link_id);
      if (nodeUuid) {
        instances.push({
          linkUuid: linkUuid ?? generateLegacyLinkUuid(sourceId, nodeUuid),
          targetUuid: nodeUuid,
          type: c.ref_type ?? 'node',
          label: c.label ?? null,
        });
      }
      return;
    }
    if (c.type === 'ref' && c.targetId) {
      if (!seenLegacyTargets.has(c.targetId)) {
        seenLegacyTargets.add(c.targetId);
        instances.push({
          linkUuid: generateLegacyLinkUuid(sourceId, c.targetId),
          targetUuid: c.targetId,
          type: 'node',
          label: c.label ?? null,
        });
      }
      return;
    }
    if (c.type === 'text' && typeof c.text === 'string') {
      let match: RegExpExecArray | null;
      REF_MARK_RE.lastIndex = 0;
      while ((match = REF_MARK_RE.exec(c.text)) !== null) {
        const targetId = match[1];
        if (!seenLegacyTargets.has(targetId)) {
          seenLegacyTargets.add(targetId);
          instances.push({
            linkUuid: generateLegacyLinkUuid(sourceId, targetId),
            targetUuid: targetId,
            type: 'node',
            label: null,
          });
        }
      }
      return;
    }
    if (Array.isArray(c.children)) {
      for (const child of c.children) walk(child);
    }
  }

  for (const child of content) walk(child);
  return instances;
}

export function rebuildNodeLinksForNode(db: Database, nodeId: string): string[] {
  const node = queryOne<{ workspace_id: string; content: string }>(
    db,
    'SELECT workspace_id, content FROM node WHERE id = ?',
    [nodeId]
  );
  if (!node) return [];

  const content = JSON.parse(node.content) as unknown[];
  const desired = extractNodeLinkInstances(content, nodeId);
  const desiredIds = new Set(desired.map((link) => link.linkUuid));

  const ts = new Date().toISOString();

  const upsert = db.prepare(
    `INSERT INTO node_link (
       id, workspace_id, source_id, target_id, type, label,
       click_count, last_navigated_at, created_at, updated_at
     ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
     ON CONFLICT(id) DO UPDATE SET
       source_id = excluded.source_id,
       target_id = excluded.target_id,
       type = excluded.type,
       label = excluded.label,
       updated_at = excluded.updated_at`
  );
  try {
    for (const link of desired) {
      upsert.run([
        link.linkUuid,
        node.workspace_id,
        nodeId,
        link.targetUuid,
        link.type,
        link.label,
        0,
        null,
        ts,
        ts,
      ]);
    }
  } finally {
    upsert.free();
  }

  const existingRows = queryAll<{ id: string }>(
    db,
    'SELECT id FROM node_link WHERE source_id = ?',
    [nodeId]
  );
  for (const row of existingRows) {
    if (!desiredIds.has(row.id)) {
      db.run('DELETE FROM node_link WHERE id = ?', [row.id]);
    }
  }

  const affectedIds = new Set<string>([nodeId]);
  for (const link of desired) {
    affectedIds.add(link.targetUuid);
  }
  return Array.from(affectedIds);
}

export function rebuildEdgesForNode(db: Database, nodeId: string): string[] {
  rebuildNodeLinksForNode(db, nodeId);

  const node = queryOne<{ workspace_id: string }>(
    db,
    'SELECT workspace_id FROM node WHERE id = ?',
    [nodeId]
  );
  if (!node) return [];

  const desiredEdges = new Map<string, { label: string | null; metadata: string }>();
  const rows = queryAll<{ target_id: string; label: string | null }>(
    db,
    'SELECT target_id, label FROM node_link WHERE source_id = ?',
    [nodeId]
  );
  for (const row of rows) {
    if (!desiredEdges.has(row.target_id)) {
      desiredEdges.set(row.target_id, {
        label: row.label,
        metadata: JSON.stringify({ label: row.label }),
      });
    }
  }

  const existingRows = queryAll<{ id: string; target_id: string; metadata: string }>(
    db,
    'SELECT id, target_id, metadata FROM edge WHERE source_id = ? AND type = ?',
    [nodeId, 'reference']
  );

  for (const row of existingRows) {
    const wanted = desiredEdges.get(row.target_id);
    if (!wanted) {
      db.run('DELETE FROM edge WHERE id = ?', [row.id]);
    } else if (row.metadata !== wanted.metadata) {
      db.run('UPDATE edge SET metadata = ? WHERE id = ?', [wanted.metadata, row.id]);
    }
    desiredEdges.delete(row.target_id);
  }

  const insert = db.prepare(
    'INSERT INTO edge (id, workspace_id, source_id, target_id, type, property_schema_id, metadata, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?)'
  );
  try {
    for (const [targetId, { metadata }] of desiredEdges) {
      insert.run([uuidv7(), node.workspace_id, nodeId, targetId, 'reference', null, metadata, new Date().toISOString()]);
    }
  } finally {
    insert.free();
  }

  const affectedIds = new Set<string>([nodeId]);
  for (const row of existingRows) {
    affectedIds.add(row.target_id);
  }
  for (const targetId of desiredEdges.keys()) {
    affectedIds.add(targetId);
  }
  return Array.from(affectedIds);
}

export function getBacklinks(db: Database, nodeId: string): string[] {
  const rows = queryAll<{ source_id: string }>(
    db,
    'SELECT DISTINCT source_id FROM node_link WHERE target_id = ? ORDER BY source_id',
    [nodeId]
  );
  return rows.map((r) => r.source_id);
}
